### PR TITLE
[4B/10] Build rendering representations through the clone lifecycle

### PR DESCRIPTION
**Pull request by ooctipus**
# Description

Follow-up to #7899 (4A). This is the clone-lifecycle step, not the SDP transport cutover.

The intended order is:

```text
Construct consumers from cfg → build one clone plan → clone representations
                            → allocate native models → initialize consumer bindings
```

## Before and after

The snippets below are shortened excerpts. Unrelated bookkeeping and arguments are omitted; they illustrate the lifecycle, not complete standalone programs.

### 1. Construct consumers before cloning; initialize them afterward

**Before:** visualizers were constructed when they were initialized, too late for their constructors to participate in clone planning.

```python
# Construction and binding happened together during visualizer initialization.
visualizer = cfg.class_type(cfg)
visualizer.initialize(self._scene_data_provider)
```

**After:** construct the configured consumers early, then bind the same objects to the completed scene later.

```python
# Before cloning: record what each viewer needs and construct it from cfg.
for cfg in self._resolve_visualizer_cfgs():
    self._render_context.clone_contexts.update(cfg.cloning_contexts)
    self._pending_visualizers.append(cfg.class_type(cfg))

# Once cloned native resources are available: bind the objects we already constructed.
for visualizer in tuple(self._pending_visualizers):
    visualizer.initialize(self._scene_data_provider)
    self._pending_visualizers.remove(visualizer)
    self._visualizers.append(visualizer)
```

Camera renderers also register during cfg processing, before the plan is built:

```python
# A camera's renderer must declare its requirements before we clone its scene.
for cfg in asset_cfgs:
    if isinstance(cfg, CameraCfg):
        sim.get_or_create_backend(cfg.renderer_cfg)
```

Renderer acquisition now uses the same cfg-based registry as native resources:

```python
# Before: a renderer-specific factory and cache, separate from native ownership.
renderer = sim.render_context.get_renderer(renderer_cfg)

# After: one simulation-owned registry constructs cfg.class_type(cfg) on a miss.
renderer = sim.get_or_create_backend(renderer_cfg)
```

`RenderContext` reads renderer entries from that registry and coordinates initialization,
stage preparation, and updates. It neither caches nor closes renderer instances. Shutdown
releases camera data, material writers, renderer bindings, visualizers, then native resources;
it does not rely on registration order to determine dependencies.

### 2. Declare required representations in cfg and route them through the same plan

**Before:** `InteractiveScene` inferred stage/model requirements from renderer and viewer type names after the sensors existed.

```python
# These flags describe what is needed, but do not route asset rows to clone contexts.
for type_name in requested_viz_types.union(cam_types):
    requires_stage, requires_model = REQUIRES_STAGE_AND_MODEL[type_name]
    self.sim.requires_usd_stage |= requires_stage
    self.sim.requires_newton_model |= requires_model
```

**After:** renderer and visualizer cfgs explicitly declare the contexts that build their representations.

```python
class NewtonWarpRendererCfg(RendererCfg):
    # This camera renderer needs a Newton representation of the declared scene.
    cloning_contexts = ("isaaclab_newton.cloner:NewtonReplicateContext",)


class NewtonVisualizerCfg(VisualizerCfg):
    # The viewer requests the same context, not a separate cloning pass.
    cloning_contexts = ("isaaclab_newton.cloner:NewtonReplicateContext",)
```

```python
# Renderer registration collects the declared requirements before planning.
self.clone_contexts.update(cfg.cloning_contexts)

# The planner routes asset rows to those contexts alongside per-asset overrides.
# Dispatch then gives every participating context the same plan (simplified).
for context_type in plan.context_rows:
    sim.clone_contexts[context_type].replicate(plan)
```

When Newton is already the physics context, its renderer/viewer requirement uses that same context. There is no separate rendering clone lifecycle. Existing stage/model flags are not all removed by this PR.

### 3. Build the Newton representation during cloning, not from a getter

**Before:** asking for the Newton model could build it by discovering and importing the completed USD stage, when another backend supplied physics.

```python
@classmethod
def get_model(cls):
    # A read could trigger scene discovery and native-model construction.
    cls._ensure_visualization_model()
    return None if cls.backend is None else cls.backend.model
```

**After:** physics and render-only Newton representations use the same import-and-replicate function. The plan supplies the routed prototypes and shared roots. Newton physics keeps its solver-specific defaults, sites, and hooks; foreign-physics rendering imports geometry without a Newton solver. There is no separate visualization-builder module.

```python
# Both physics and rendering enter this same construction path.
return _replicate_newton(
    self._sim.stage, plan, plan.context_rows[type(self)], self._sim, up_axis=self.up_axis
)
```

```python
# Inside the shared path: choose physics defaults or a plain geometry builder.
create_builder = partial(manager_cls.create_builder if simulation else ModelBuilder, up_axis=up_axis)

# Import each prototype once, then expand the same clone mapping in either mode.
source_builders = build_source_builders(stage, sources, create_builder, schema_resolvers, ...)
replicate_builder_mapping(...)  # Uses the plan's routed mask, positions, env IDs, and destinations.

# For foreign physics, allocate the resulting model only after physics is ready.
backend_cfg = NewtonBackendCfg(
    builder=builder, device=sim.device, num_envs=len(plan.env_ids), simulation=False
)
sim.physics_manager.register_callback(
    partial(NewtonManager._initialize_visualization_model, backend_cfg, geometry),
    PhysicsEvent.PHYSICS_READY,
    name="newton_visualization_model",
    wrap_weak_ref=False,
)
```

```python
# Inside that readiness callback: use the native-resource registry introduced in 4A.
NewtonManager.backend = sim.get_or_create_backend(cfg)


@classmethod
def get_model(cls):
    # Reading the model no longer constructs anything or scans the stage.
    return None if cls.backend is None else cls.backend.model
```

This removes late full-stage reconstruction from this rendering path. It does **not** eliminate every geometry scan: deformable metadata is still inspected within the declared source/shared roots.

### 4. Make standalone demos declare what belongs to their scene

**Before:** standalone demos could author geometry without a plan, relying on a Newton-backed viewer to discover it later.

```python
scene_entities, scene_origins = design_scene()
sim.reset()  # A viewer could reconstruct its representation from the finished stage.
```

**After:** the quadruped demo declares its roots before authoring and asks the clone lifecycle to build the required representations before reset.

```python
# This demo authors independent roots, so it uses a global-only plan.
global_paths = (
    "/World/defaultGroundPlane", "/World/Light",
    *(f"/World/Origin{i}" for i in range(1, 8)),
)
plan = cloner.make_clone_plan((), 1, 0.0, global_paths=global_paths)
sim.set_clone_plan(plan)

# Author the declared scene, then build its rendering representations.
scene_entities, scene_origins = design_scene()
cloner.replicate(plan, replicate_physics=False)  # Keep the existing native physics setup.
sim.reset()
```

The same minimal migration covers the affected standalone demos and URDF/MJCF previews. `InteractiveScene` handles the ordering automatically; its callers do not need this explicit ceremony.

## Scope and size

No new production classes or files. Production **−274 LOC**, tests **−99 LOC** against develop `eaab09fd0` after the conflict-resolution merge. The builder-consolidation follow-up alone removed **122 production lines**, including `visualization_builder.py` and two redundant helper layers. The renderer-registry follow-up removed another **21 production lines** and **93 test lines** while adding teardown-order and registration regressions. Config-keyed sharing now also owns renderer instances; supported native-model sharing cases are unchanged. Independently configured Newton models, singleton query routing, and SDP conversion/copy changes remain outside this PR. Performance samples below retain their explicitly pinned historical baselines; they were not rerun for this merge.

## Initial develop comparison

Baseline: merged develop `c90c3a1097a0a669a823b880b1c8d0654d7fd267`. GPU 0, RTX 5090, `Isaac-Lift-KukaAllegro-Camera`, 4096 environments, seed 0, no interactive visualizer. One warm sample each, same installed dependencies/cache and resolved workload configuration. Startup includes imports, construction, reset, and first step. Runtime discards 25 steps, then measures 200 synchronized full environment steps, including actions, camera observations, and episode resets.

| Physics → renderer | Startup 4B / develop | Runtime ms/step 4B / develop | Environment FPS 4B / develop |
| --- | ---: | ---: | ---: |
| Newton MJWarp → Newton | 28.34 / 28.33 s | 46.31 / 45.94 | 88,457 / 89,159 |
| OVPhysX → Newton | 62.94 / 62.95 s | 163.02 / 160.83 | 25,126 / 25,468 |

These samples preceded the builder-consolidation follow-up. Startup was effectively unchanged; runtime was 0.8% and 1.4% slower respectively, which a single pair cannot establish as significant. This measures incremental 4B, not the later full-refactor optimization stack. All 4096 camera outputs were finite and nonconstant. The separate follow-up check below compares consolidation against its immediate parent; do not combine timings across the two sessions.

### Builder-consolidation follow-up check

GPU 0, same 4096-environment workload, seed, dependencies, `_isaac_sim` launcher, 25 discarded steps and 200 measured steps. One warm sample per revision/path. Before is this PR's previous head `2f75d462a8`; after is the consolidated pipeline committed as `b39adbd4d`. Resolved configurations were byte-identical. Earlier temporary-worktree measurements without the matching Isaac Sim launcher were discarded.

| Physics → renderer | Startup after / before | Runtime ms/step after / before | Environment FPS after / before |
| --- | ---: | ---: | ---: |
| Newton MJWarp → Newton | 28.73 / 29.02 s | 50.32 / 48.14 | 81,395 / 85,090 |
| OVPhysX → Newton | 68.33 / 65.56 s | 174.20 / 163.59 | 23,513 / 25,038 |

These are **not a demonstrated no-regression result**: startup changed −1.0% / +4.2% and runtime +4.5% / +6.5%. There was no competing GPU 0 compute job, but host CPU activity was not isolated. One pair cannot establish whether those differences are repeatable. This follow-up changes construction only, not the stepping/render transport loop. All 4096 camera outputs passed the finite/nonconstant checks.

## Validation

- Conflict resolution against develop `eaab09fd0`: preserved upstream MPM particle-range tracking in the shared clone path. 114 focused tests plus four subtests passed, including GPU 0 MPM initialization/reset and OVRTX cleanup checks. Formatting/changelog validation passed against the updated base.
- Renderer-registry follow-up: 163 focused CPU tests and 11 profiler-target checks passed. Five regression cases failed when the prior registry methods were substituted: pre-construction global-setting validation and renderer-before-native teardown in both creation orders, with and without cleanup failures.
- Builder-consolidation follow-up: 172 focused CPU tests (plus four subtests) and four solver-specific import cases passed, covering cloning, source geometry, renderer registration, visualizer lifecycle, native ownership, and SDP copy/remap behavior.
- Regression checks failed on the pre-consolidation commit and passed here: omitted clone positions and excluded deformable rows beneath an overlapping shared root. Public direct replication kept its signature.
- Standalone script contract checks passed (launch matrix not run in full).
- GPU 0: Newton/OVPhysX → Newton camera smokes; OVPhysX → OVRTX camera smoke; OVPhysX camera + headless Newton GL shared one native model across an ordinary reset; migrated Isaac PhysX marker and OVPhysX deformable demos reached readiness and stepped.
- After registry consolidation, repeated 16-environment GPU 0 Kuka camera checks for Newton → Newton, OVPhysX → OVRTX, and OVPhysX → Newton camera + headless Newton GL. All passed, including valid images, model sharing/reset, and teardown. These are correctness smokes, not new matched performance samples.
- `uv run --no-sync isaaclab -f` passed, with changelog validation pinned to the actual upstream baseline. After registry consolidation, the full isolated documentation build (`uv run --no-sync isaaclab -d`, using the branch's docs-enabled `dev` extra) passed with warnings treated as errors.

The pre-existing forced native-rewarm issue, where some consumers retain old model bindings after STOP/recreation, is not claimed fixed here. Ordinary reset and first initialization are covered.

## Type of change

- Refactor and lifecycle correction.
- Breaking behavior for external standalone foreign-physics rendering workflows that never declared a plan: declare source/shared roots before cloning. Maintained affected demos are migrated; the cloning guide includes the standalone pattern.
- Removed `RenderContext.get_renderer`: use `sim.get_or_create_backend(renderer_cfg)`. Use `sim.render_context` for orchestration; its constructor now takes the shared registry.

## Release backport

- [x]  Backport this pull request to the active release branch after it merges into `develop`

## Checklist

- [x] Contribution guidelines read.
- [x] Pre-commit checks passed.
- [x] Documentation and per-package changelog fragments updated.
- [x] Focused tests and GPU smoke checks completed.
- [x] Contributor already listed.

### CODE REVIEW

**Comment by ooctipus** [comment]
run-ci

File: source/isaaclab_ov/isaaclab_ov/renderers/ovrtx_renderer_cfg.py [hunk 1]
```diff
@@ -53,6 +53,9 @@ class OVRTXRendererCfg(RendererCfg):
     renderer_type: str = "ovrtx"
     """Type identifier for OVRTX renderer."""
 
+    cloning_contexts: tuple[type | str, ...] = ("isaaclab_newton.cloner:NewtonReplicateContext",)
```
Comment (StafaH) [inline source/isaaclab_ov/isaaclab_ov/renderers/ovrtx_renderer_cfg.py:56]: ovrtx uses newton replicate context?

File: source/isaaclab_newton/isaaclab_newton/cloner/replicate.py [hunk 2]
```diff
@@ -242,7 +244,21 @@ def replicate(self, plan: ClonePlan) -> tuple[ModelBuilder, object, dict]:
             raise ValueError("ClonePlan.env_ids is required for replication.")
         rows = plan.context_rows[type(self)]
         cfg = self._sim.cfg.physics
-        load_visual_shapes = cfg.load_visual_shapes if isinstance(cfg, NewtonCfg) else None
+        if not isinstance(cfg, NewtonCfg):
+            builder, geometry = build_visualization_builder_from_plan(
```
Comment (StafaH) [inline source/isaaclab_newton/isaaclab_newton/cloner/replicate.py]: I thought we are not doing visualization builder anymore (and just keeping one builder for newton)?

File: source/isaaclab/isaaclab/cloner/clone_plan.py [hunk 3]
```diff
@@ -297,10 +311,13 @@ def make_clone_plan(
 
     cfgs = tuple(cfgs)
     global_paths = _minimal_roots(global_paths)
+    sim = sim_utils.SimulationContext.instance()
```
Comment (StafaH) [inline source/isaaclab/isaaclab/cloner/clone_plan.py:314]: I wonder if we should pass sim as a argument (so that we can begin to think of sim as not a singleton global). This will help in future when the singleton is removed?

File: source/isaaclab/isaaclab/sim/simulation_context.py [hunk 4]
```diff
@@ -245,7 +245,9 @@ def __init__(self, cfg: SimulationCfg | None = None):
             order=5,
         )
 
-        type(self)._instance = self  # Mark as valid singleton only after successful init
+        # Publish the context before configured consumers register their clone requirements.
+        type(self)._instance = self
+        self._create_visualizers()
```
Comment (StafaH) [inline source/isaaclab/isaaclab/sim/simulation_context.py:252]: **Major:** Roll back the singleton when pre-clone consumer construction fails

`_instance` is published immediately before `_create_visualizers()`, and that method can now import and construct optional backends. If any constructor raises, `SimulationContext(...)` fails but `SimulationContext.instance()` retains the partial object; subsequent construction cannot retry (and current `develop` now explicitly raises whenever an instance already exists). Construct into temporary state and publish only after success, or wrap this phase in rollback that closes created consumers/resources and clears `_instance`. Add a regression test that fails one visualizer constructor, verifies `instance()` is `None`, and then successfully constructs a replacement context.

File: source/isaaclab/isaaclab/sim/simulation_context.py [hunk 5]
```diff
@@ -633,79 +635,31 @@ def _resolve_visualizer_cfgs(self) -> list[Any]:
...
-                requires_stage, requires_model = REQUIRES_STAGE_AND_MODEL[config.visualizer_type]
+    def _create_visualizers(self) -> None:
+        """Construct cfg-owned consumers and publish their requirements before scene cloning."""
+        for cfg in self._resolve_visualizer_cfgs():
+            if cfg.visualizer_type is not None:
+                requires_stage, requires_model = REQUIRES_STAGE_AND_MODEL[cfg.visualizer_type]
                 self.requires_usd_stage |= requires_stage
                 self.requires_newton_model |= requires_model
+            self._render_context.clone_contexts.update(cfg.cloning_contexts)
+            self._pending_visualizers.append(cfg.class_type(cfg))
```
Comment (StafaH) [inline source/isaaclab/isaaclab/sim/simulation_context.py:651]: **Major:** Preserve an explicit required-versus-optional visualizer failure policy

Construction and later initialization now propagate for every resolved visualizer. Previously only an explicitly CLI-requested visualizer was fatal; configured defaults and XR auto-injection logged and degraded gracefully. An unavailable optional backend can therefore abort `SimulationContext.__init__()` or `reset()`, and the new changelog does not disclose this compatibility change. Define required/optional semantics explicitly (at minimum keep XR auto-injection best-effort), preserve the old boundary or document the break with migration guidance, and wrap failures with the visualizer type plus install hint.

File: source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py [hunk 6]
```diff
@@ -25,148 +23,66 @@
...
-        destinations=destinations,
-        env_ids=env_ids,
-    )
-    shadow_entities, registry_groups = add_shadow_deformables_to_builder(
-        builder, stage, env_paths, device=device, entries=deformable_entries, clone_plan=clone_plan
+        destinations=tuple(plan.destinations[row] for row in rows),
+        env_ids=plan.env_ids,
     )
-    return builder, (shadow_entities, registry_groups)
+    return builder, add_shadow_deformables_to_builder(builder, stage, entries, plan, device=device)
```
Comment (StafaH) [inline source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py]: **Major:** Restrict deformable expansion to the rows routed to this context

Rigid geometry uses `rows`, but deformable expansion receives the complete plan. This is observable when a declared global root contains clone sources (an overlap this PR's own builder test supports): discovery sees a deformable under a source excluded from the Newton context, then `_expand_clone_plan_deformable_entries()` walks every plan row and clones it anyway. The resulting render model violates `context_rows`. Pass a routed sub-plan/row set through deformable grouping and expansion, and add an overlap test with one included and one excluded source row.

File: source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py [hunk 7]
```diff
@@ -25,148 +23,66 @@
...
     builder.add_builder(global_builder)
+    quaternions = np.zeros((len(plan.env_ids), 4), dtype=np.float32)
+    quaternions[:, 3] = 1.0
     replicate_builder_mapping(
         builder=builder,
         sources=sources,
-        mapping=mapping,
-        positions=positions,
+        mapping=plan.clone_mask[list(rows)],
+        positions=plan.positions,
```
Comment (StafaH) [inline source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py]: **Major:** Honor the optional `ClonePlan.positions` contract

`ClonePlan.positions` is public and optional, and the other clone contexts accept `None`; however this new path forwards `None` to `replicate_builder_mapping()`, which immediately calls `positions.astype(...)`. Deformable expansion also indexes `clone_plan.positions`. A valid hand-built plan therefore crashes only when a foreign-physics Newton renderer/visualizer is selected. Normalize missing positions to a zero array (as Newton physics replication already does), or make positions required for all executable plans and validate that consistently at dispatch with a clear error.

File: source/isaaclab/isaaclab/cloner/clone_plan.py [hunk 8]
```diff
@@ -229,9 +230,22 @@ def _context_rows(
...
-    rows_by_context: dict[type[object], set[int]] = (
-        {} if physics_context is None or not global_paths else {physics_context: set()}
-    )
+    render_contexts = {
+        string_to_callable(context) if isinstance(context, str) else context
+        for context in sim.render_context.clone_contexts
+    }
+    for context_type in render_contexts:
+        if context_type not in sim.clone_contexts:
+            sim.clone_contexts[context_type] = context_type(sim)
```
Comment (StafaH) [inline source/isaaclab/isaaclab/cloner/clone_plan.py:239]: **Major:** Define one clone-context factory contract before exposing it in configs

The new public `RendererCfg.cloning_contexts` / `VisualizerCfg.cloning_contexts` accepts arbitrary context classes, but missing contexts are instantiated as `context_type(sim)`. Existing built-ins do not share that contract: Newton and OvPhysX take a `SimulationContext`, while USD and PhysX take a stage. For example, a renderer that declares `PhysxReplicateContext` under another physics backend fails inside its constructor with the wrong object. Route creation through an explicit simulation-owned registry/factory (or standardize every context constructor) and document/type the protocol; do not infer construction from a raw class reference.

File: source/isaaclab/isaaclab/renderers/render_context.py [hunk 9]
```diff
@@ -140,6 +143,7 @@ def get_renderer(self, cfg: RendererCfg) -> BaseRenderer:
             raise RuntimeError("Renderers must be registered before rendering consumers are finalized.")
         new_renderer = cfg.class_type(cfg)
         self._renderer_entries.append((cfg, new_renderer))
+        self.clone_contexts.update(cfg.cloning_contexts)
```
Comment (StafaH) [inline source/isaaclab/isaaclab/renderers/render_context.py:131]: **Minor:** Reject representation requirements registered after the plan is frozen

`get_renderer()` continues to accept a new renderer and mutate `clone_contexts` after `SimulationContext` already owns a `ClonePlan`. That renderer's context is absent from `plan.context_rows`, so it is never replicated; with lazy Newton model construction removed, users get a missing/stale model failure much later during camera initialization or rendering. Freeze representation registration when the plan is built and raise an actionable error here when a newly requested context is not present (for example: register this camera/renderer before cloning).

**Review by StafaH — COMMENTED** [review]
Some AI review comments as well

File: source/isaaclab_newton/isaaclab_newton/cloner/replicate.py (hunk 2 above)
Comment (ooctipus) [reply source/isaaclab_newton/isaaclab_newton/cloner/replicate.py]: Consolidated in b39adbd4d. Physics and render-only Newton construction now share one plan-driven import/replication path; visualization_builder.py and the forwarding helper are removed. Physics retains solver-specific defaults, sites, and hooks; render-only models retain compact collision-filter storage and PHYSICS_READY allocation. This follow-up removes 122 production lines. Focused tests and GPU camera/viewer sharing checks passed; the description includes the measured startup/runtime results and their limitations.

File: source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py (hunk 6 above)
Comment (ooctipus) [reply source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py]: Fixed in b39adbd4d: routed rows now reach deformable expansion. The overlapping-global-root regression excludes the unselected source and passes here; it fails against the previous head.

File: source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py (hunk 7 above)
Comment (ooctipus) [reply source/isaaclab_newton/isaaclab_newton/physics/visualization_builder.py]: Fixed in b39adbd4d without changing the public optional-positions contract. The shared construction path normalizes missing positions, and shadow deformables use zero offsets. Existing clone and deformable tests now cover None; the previous head fails the full-context regression.

**Comment by ooctipus** [comment]
run-ci

File: source/isaaclab_ov/isaaclab_ov/renderers/ovrtx_renderer_cfg.py (hunk 1 above)
Comment (ooctipus) [reply source/isaaclab_ov/isaaclab_ov/renderers/ovrtx_renderer_cfg.py:56]: this is be gone in 4c

File: source/isaaclab/isaaclab/cloner/clone_plan.py (hunk 3 above)
Comment (ooctipus) [reply source/isaaclab/isaaclab/cloner/clone_plan.py:314]: I want to do a single pr that remove the singelton, but not today!

**Comment by ooctipus** [comment]
run-ci

**Comment by ooctipus** [comment]
run-ci

**Review by StafaH — APPROVED** [review]
